crates/miden-objects/src/account/component/template/storage/placeholder.rs

```diff
@@ -60,142 +65,403 @@ impl StoragePlaceholder {
     /// This method returns an error if:
     /// - Any segment (or the whole key) is empty.
     /// - Any segment contains invalid characters.
-    pub fn new(key: impl Into<String>) -> Result<Self, StoragePlaceholderError> {
-        let key: String = key.into();
-        Self::validate(&key)?;
-        Ok(Self { key })
+    pub fn new(base: impl Into<String>) -> Result<Self, StorageValueNameError> {
+        let base: String = base.into();
+        for segment in base.split('.') {
+            Self::validate_segment(segment)?;
+        }
+        Ok(Self { fully_qualified_name: base })
+    }
+
+    /// Adds a suffix to the storage value name, separated with a dot.
+    #[must_use]
+    pub fn with_suffix(self, suffix: &StorageValueName) -> StorageValueName {
+        let mut key = self;
+        if !suffix.inner().is_empty() {
+            if !key.inner().is_empty() {
+                key.fully_qualified_name.push('.');
+            }
+            key.fully_qualified_name.push_str(&suffix.to_string());
+        }
+
+        key
     }
 
-    /// Returns the key name
     pub fn inner(&self) -> &str {
-        &self.key
+        &self.fully_qualified_name
     }
 
-    /// Checks if the given string is a valid key.
-    /// A storage placeholder is valid if it's made of one or more segments that are non-empty
-    /// alphanumeric strings.
-    fn validate(key: &str) -> Result<(), StoragePlaceholderError> {
-        if key.is_empty() {
-            return Err(StoragePlaceholderError::EmptyKey);
+    fn validate_segment(segment: &str) -> Result<(), StorageValueNameError> {
+        if segment.is_empty() {
+            return Err(StorageValueNameError::EmptySegment);
+        }
+        if !segment.chars().all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-') {
+            return Err(StorageValueNameError::InvalidCharacter {
+                part: segment.to_string(),
+                character: segment
+                    .chars()
+                    .find(|c| !(c.is_ascii_alphanumeric() || *c == '_' || *c == '-'))
+                    .unwrap(),
+            });
         }
+        Ok(())
+    }
+}
 
-        for segment in key.split('.') {
-            if segment.is_empty() {
-                return Err(StoragePlaceholderError::EmptyKey);
-            }
+impl Display for StorageValueName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.inner())
+    }
+}
 
-            for c in segment.chars() {
-                if !(c.is_ascii_alphanumeric() || c == '_' || c == '-') {
-                    return Err(StoragePlaceholderError::InvalidChar(key.into(), c));
-                }
-            }
-        }
+impl Serializable for StorageValueName {
```

Comment:
Do we actually need to implement `Serializable` for all the template-related types? Where would we use the binary representations? Looks like they are currently unused.

Reply:
Well, the original intent for templates was that they would contain storage metadata + library, and these would be serialized into a single "package" file. This means that serialization propagates down to most/all structs. Additionally, `serde` and `toml` are only enabled when `std` is set, so this also provides a `no_std` way of serializing.